[pytorch][PR] [pt1][quant] Add slicing and int_repr() to QTensor

jerryzh168 opened the pull request:
Stack:
&nbsp;&nbsp;&nbsp;&nbsp;:black_circle:&nbsp; **#19296 [pytorch][PR] [pt1][quant] Add slicing and int_repr() to QTensor**&nbsp;&nbsp;[:green_heart:](https://our.internmc.facebook.com/intern/diff/D14756833/)
&nbsp;&nbsp;&nbsp;&nbsp;:white_circle:&nbsp; #18960 [pt1][quant] Add empty_quantized&nbsp;&nbsp;[:yellow_heart:](https://our.internmc.facebook.com/intern/diff/D14810261/)

Stack:
&nbsp;&nbsp;&nbsp;&nbsp;:black_circle:&nbsp; **#19296 [pt1][quant] Add slicing and int_repr() to QTensor**&nbsp;&nbsp;[:yellow_heart:](https://our.intern.facebook.com/intern/diff/D14756833/)
&nbsp;&nbsp;&nbsp;&nbsp;:white_circle:&nbsp; #18960 [pt1][quant] Add empty_quantized&nbsp;&nbsp;[:yellow_heart:](https://our.intern.facebook.com/intern/diff/D14810261/)
&nbsp;&nbsp;&nbsp;&nbsp;:white_circle:&nbsp; #19312 Use the QTensor with QReLU&nbsp;&nbsp;[:yellow_heart:](https://our.intern.facebook.com/intern/diff/D14819460/)
&nbsp;&nbsp;&nbsp;&nbsp;:white_circle:&nbsp; #19319 [RFC] Quantized SumRelu&nbsp;&nbsp;[:yellow_heart:](https://our.intern.facebook.com/intern/diff/D14866442/)

Methods added to pytorch python frontend:
- int_repr() returns a CPUByte Tensor which copies the data of QTensor.
- Added as_strided for QTensorImpl which provides support for slicing a QTensor(see test_torch.py)
Pull Request resolved: https://github.com/pytorch/pytorch/pull/19296

Differential Revision: [D14756833](https://our.internmc.facebook.com/intern/diff/D14756833/)

dzhulgakov commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -204,6 +204,10 @@
 
 - func: as_strided(Tensor(a) self, int[] size, int[] stride, int? storage_offset=None) -> Tensor(a)
   variants: function, method
+  dispatch:
```
looks like adding dispatch here broke dispatch for bool tensors (our codegen is crazy). Probably the fix is to add
```
  cpu_half: True
  cpu_bool: True
  cuda_bool: True
```
here. @gchanan @smessmer to double check

jerryzh168 committed:
Add slicing and int_repr() to QTensor (#19296)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/19296

Pull Request resolved: https://github.com/pytorch/pytorch/pull/18814

Methods added to pytorch python frontend:
- int_repr() returns a CPUByte Tensor which copies the data of QTensor.
- Added as_strided for QTensorImpl which provides support for slicing a QTensor(see test_torch.py)

Differential Revision: D14756833

fbshipit-source-id: 752e42ebb55826ab8780bb7d3215f40c81737874
